packages/instantsearch-ui-components/src/components/Carousel.tsx

```diff
@@ -0,0 +1,222 @@
+/** @jsx createElement */
+import { cx } from '../lib';
+
+import type {
+  ComponentProps,
+  RecommendItemComponentProps,
+  RecordWithObjectID,
+  Renderer,
+} from '../types';
+
+export type CarouselProps<
+  TObject,
+  TComponentProps extends Record<string, unknown> = Record<string, unknown>
+> = ComponentProps<'div'> & {
+  items: Array<RecordWithObjectID<TObject>>;
+  itemComponent: (
+    props: RecommendItemComponentProps<RecordWithObjectID<TObject>> &
+      TComponentProps
+  ) => JSX.Element;
+  classNames?: Partial<CarouselClassNames>;
+  translations?: Partial<CarouselTranslations>;
+};
+
+export type CarouselClassNames = {
+  /**
+   * Class names to apply to the root element
+   */
+  root: string | string[];
+  /**
+   * Class names to apply to the list element
+   */
+  list: string | string[];
+  /**
+   * Class names to apply to each item element
+   */
+  item: string | string[];
+  /**
+   * Class names to apply to both navigation elements
+   */
+  navigation: string | string[];
+  /**
+   * Class names to apply to the next navigation element
+   */
+  navigationNext: string | string[];
+  /**
+   * Class names to apply to the previous navigation element
+   */
+  navigationPrevious: string | string[];
+};
+
+export type CarouselTranslations = {
+  /**
+   * The label of the next navigation element
+   */
+  nextButtonLabel: string;
+  /**
+   * The title of the next navigation element
+   */
+  nextButtonTitle: string;
+  /**
+   * The label of the previous navigation element
+   */
+  previousButtonLabel: string;
+  /**
+   * The title of the previous navigation element
+   */
+  previousButtonTitle: string;
+  /**
+   * The label of the carousel
+   */
+  listLabel: string;
+};
+
+let lastCarouselId = 0;
+
+function generateCarouselId() {
+  return `ais-Carousel-${lastCarouselId++}`;
+}
+
+export function createCarouselComponent({ createElement, useRef }: Renderer) {
+  // This prevents the SVG buttons references from being hoisted
+  // during the build, where `createElement` is not available.
+  const buttonPathCommonProps: JSX.IntrinsicElements['path'] = {
+    fillRule: 'evenodd',
+    clipRule: 'evenodd',
+    fill: 'currentColor',
+  };
```

Comment:
📝 I could not prevent Babel from hoisting simple JSX references outside of `createCarouselComponent`, so I resorted to this "hack"

Reply:
wonder if disabling https://babeljs.io/docs/babel-plugin-transform-react-constant-elements could have helped?

Reply:
It does, thanks for pointing it out. I could remove the plugin for all ui-components, we're basically expecting every element to be rendered through `createElement` which makes this opt uninteresting in this package. Wdyt?

Reply:
yes, I think we should do that chagne

Reply:
Done in [3b86cbf](https://github.com/algolia/instantsearch/pull/6289/commits/3b86cbf54c6daa210f8f4d3af624f15ec0dde1a8).